compiler/rustc_codegen_llvm/src/back/archive.rs

```diff
@@ -351,3 +403,26 @@ impl<'a> LlvmArchiveBuilder<'a> {
 fn string_to_io_error(s: String) -> io::Error {
     io::Error::new(io::ErrorKind::Other, format!("bad archive: {}", s))
 }
+
+fn find_dlltool(sess: &Session) -> OsString {
+    // When cross-compiling first try binary prefixed with target triple
+    if sess.host.llvm_target != sess.target.llvm_target {
+        let prefixed_dlltool = if sess.target.arch == "x86" {
+            "i686-w64-mingw32-dlltool"
+        } else {
+            "x86_64-w64-mingw32-dlltool"
+        };
+        let prefixed_dlltool = if cfg!(windows) {
+            [prefixed_dlltool, "exe"].concat()
+        } else {
+            prefixed_dlltool.to_string()
+        };
+        for dir in env::split_paths(&env::var_os("PATH").unwrap_or_default()) {
```

Comment:
This should also either verify that the binary is executable at all in some way or provide a way to override this detection. Right now if the system is borked enough to have a non-working `i686-w64-mingw32-dlltool` but a working `dlltool`, then there is fairly little the user user can do to work around the problem.

Reply:
TBH the more I think about this approach the more hacky it seems.